src/lib/log/audit/audit.go

```diff
@@ -0,0 +1,148 @@
+type LoggerManager struct {
+	currentEndpoint string
+	cache           cache.Cache
+	auditLogger     *log.Logger
+}
+
+// Init redirect the audit log to the forward endpoint
+func (a *LoggerManager) Init(ctx context.Context, logEndpoint string) {
+	a.cache = cache.Default()
+	a.currentEndpoint = logEndpoint
+	al, err := syslog.Dial("tcp", logEndpoint,
+		syslog.LOG_INFO, "audit")
+	if err != nil {
+		log.Errorf("failed to create audit log, error %v", err)
+		a.SetEndpointStatus(ctx, logEndpoint, false)
+		return
+	}
+	a.auditLogger = log.New(al, log.NewTextFormatter(), log.InfoLevel, 3)
+	a.SetEndpointStatus(ctx, logEndpoint, true)
+}
+
+// DefaultLogger ...
+func (a *LoggerManager) DefaultLogger(ctx context.Context) *log.Logger {
+	endpoint := config.AuditLogForwardEndpoint(ctx)
+	if a.currentEndpoint != endpoint {
+		a.Init(ctx, endpoint)
+	}
+	if a.IsActive(ctx, endpoint) {
+		return a.auditLogger
+	}
+	return log.DefaultLogger()
+}
+
+// IsActive fetch the endpoint status from cache
+func (a *LoggerManager) IsActive(ctx context.Context, address string) bool {
+	var val string
+	cacheKey := fmt.Sprintf("endpoint:%v", address)
+	err := a.cache.Fetch(ctx, cacheKey, &val)
+	if err != nil {
+		return false
+	}
+	return strings.EqualFold(val, "true")
+}
+
+// RefreshEndpointStatus check the endpoint liveness and store it in cache
+func (a *LoggerManager) RefreshEndpointStatus(ctx context.Context) {
+	address := config.AuditLogForwardEndpoint(ctx)
+	if len(address) == 0 {
+		return
+	}
+	if a.currentEndpoint == address {
+		a.SetEndpointStatus(ctx, address, CheckEndpointActive(address))
+		return
+	}
+
+	// If there are multiple harbor-core instances, one of the harbor-core change the endpoint config
+	// other harbor-core instances should switch audit log to the new endpoint
+	log.Info("Refresh audit log forward endpoint")
+	a.Init(ctx, address)
+}
+
+// CheckEndpointActive check the liveness of the endpoint
+func CheckEndpointActive(address string) bool {
+	al, err := syslog.Dial("tcp", address,
+		syslog.LOG_INFO, "audit")
+	if al != nil {
+		defer al.Close()
+	}
+	if err != nil {
+		log.Errorf("failed to connect to audit log endpoint, error %v", err)
+		return false
+	}
+	return true
+}
+
+// SetEndpointStatus save the log forward endpoint status in cache
+func (a *LoggerManager) SetEndpointStatus(ctx context.Context, address string, active bool) {
+	cacheKey := fmt.Sprintf("endpoint:%v", address)
+	val := "false"
+	if active {
+		val = "true"
+	}
+	err := a.cache.Save(ctx, cacheKey, val, cacheInterval)
+	if err != nil {
+		log.Errorf("failed to cache endpoint status, error %v, the audit log may log to standard log", err)
+	}
+	a.currentEndpoint = address
+}
+
+// StartEndpointHealthCheck refresh the endpoint status periodically
+func StartEndpointHealthCheck(ctx context.Context, closing, done chan struct{}) {
+	auditEP := config.AuditLogForwardEndpoint(ctx)
+	LogMgr.Init(ctx, auditEP)
```

Comment:
the init here is in a gorountine, there is no guarantee that init will be executed successfully before it is used.